### PR TITLE
#376: Add WildFly-Swarm health check enricher. Not yet working becaus…

### PR DESCRIPTION
**Pull request by davsclaus**
…e fmp doesnt call it.

### CODE REVIEW

**Comment by davsclaus** [comment]
Hey @rhuss I started on a health check enricher for wildfly-swarm. 

This comes OOTB in swarm if you have a dependency on that monitor dependency.

But our quickstart
https://github.com/fabric8-quickstarts/swarm-camel

Does not make fmp generate the livness and readiness probes in the generate k8s resource files.

I debugged the code a bit and in 
https://github.com/fabric8io/fabric8-maven-plugin/blob/26126525719c363c5e608d8d15dd4dad58795e49/enricher/api/src/main/java/io/fabric8/maven/enricher/api/AbstractHealthCheckEnricher.java#L22-L22

Then item is never an instance of Deployment and the code below is not triggered that adds those probes. 

I wonder if you could take a look?

**Comment by rhuss** [comment]
Intersting, because afaik there _should_ be a `Deployment` object egernated automatically by the `DefaultControllerEnricher`. Of course the Deployment is only available if the DefaultControllerEnricher is running before the HealtCheckEnricher. 

I will check that.

**Comment by rhuss** [comment]
In fact, the enricher for the health-checks kicks in to early (as it parts of the `f8` enricher in contrast to the `fmp` standard enrichers).

The solution is to add a high order value to the health check enrichers.

However, the _real_ solution is to separate the generation of default resources (like the `Deployment` in this case) with the enrichment process (used by the health-checl). See #678 for more details on this.

I will submit a PR for the order issue.

**Comment by rhuss** [comment]
@davsclaus could you try to add a high order number to the healtcheck enrichers in https://github.com/rhuss/fabric8-maven-plugin/blob/e9bc60ab870e8137a6697883b22e9cc536b1d8ae/enricher/fabric8/src/main/resources/META-INF/fabric8/enricher-default#L9-L8 ? (like for the ExposeController) and check whether this works for you ?

I'm currently working on a more general solution to explicit define the order instead on relying of an order when merging enrichers from different files.

**Comment by davsclaus** [comment]
Thanks @rhuss that worked with ,1000, eg I see now
```
[INFO] --- fabric8-maven-plugin:3.2-SNAPSHOT:resource (default) @ swarm-camel ---
[INFO] F8: Running in Kubernetes mode
[INFO] F8: Running generator wildfly-swarm
2016-12-05 11:23:47 INFO  Version:30 - HV000001: Hibernate Validator 5.2.4.Final
[INFO] F8: fmp-controller: Adding a default Deployment
[INFO] F8: fmp-service: Adding a default Service with ports [8080]
[INFO] F8: wildfly-swarm-health-check: Adding readiness probe on port 8080, path='/health', scheme='HTTP', with initial delay 10 seconds
[INFO] F8: wildfly-swarm-health-check: Adding liveness probe on port 8080, path='/health', scheme='HTTP', with initial delay 180 seconds
```

And yay the readiness only signal its ready to k8s when swarm is really started after 31 seconds
```
NAME                           READY     STATUS    RESTARTS   AGE
swarm-camel-2856146500-ng5yy   0/1       Running   0          5s
davsclaus:/Users/davsclaus/$ kubectl get pods
NAME                           READY     STATUS    RESTARTS   AGE
swarm-camel-2856146500-ng5yy   0/1       Running   0          8s
davsclaus:/Users/davsclaus/$ kubectl get pods
NAME                           READY     STATUS    RESTARTS   AGE
swarm-camel-2856146500-ng5yy   0/1       Running   0          8s
davsclaus:/Users/davsclaus/$ kubectl get pods
NAME                           READY     STATUS    RESTARTS   AGE
swarm-camel-2856146500-ng5yy   0/1       Running   0          13s
davsclaus:/Users/davsclaus/$ kubectl get pods
NAME                           READY     STATUS    RESTARTS   AGE
swarm-camel-2856146500-ng5yy   0/1       Running   0          14s
davsclaus:/Users/davsclaus/$ kubectl get pods
NAME                           READY     STATUS    RESTARTS   AGE
swarm-camel-2856146500-ng5yy   0/1       Running   0          21s
davsclaus:/Users/davsclaus/$ kubectl get pods
NAME                           READY     STATUS    RESTARTS   AGE
swarm-camel-2856146500-ng5yy   1/1       Running   0          31s
```

Before the pod was started asap and it would of course fail to service the traffic.

File: enricher/fabric8/src/main/java/io/fabric8/maven/enricher/fabric8/WildFlySwarmHealthCheckEnricher.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    protected Probe getLivenessProbe() {
+        Probe probe = discoverWildFlySwarmHealthCheck(180);
+        return probe;
+    }
+
+    private Probe discoverWildFlySwarmHealthCheck(int initialDelay) {
+        if (hasDependency(this.getProject(), "org.wildfly.swarm", "monitor")) {
+            // TODO: wildfly-swarm does not yet have a standard configuration file
+            Integer port = DEFAULT_HEALTH_PORT;
+            String scheme = SCHEME_HTTP;
```
Comment (jimmidyson) [inline enricher/fabric8/src/main/java/io/fabric8/maven/enricher/fabric8/WildFlySwarmHealthCheckEnricher.java]: Can we make the scheme configurable via property please?

File: enricher/fabric8/src/main/java/io/fabric8/maven/enricher/fabric8/WildFlySwarmHealthCheckEnricher.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+import io.fabric8.maven.enricher.api.EnricherContext;
+
+import static io.fabric8.maven.core.util.MavenUtil.hasDependency;
+
+/**
+ * Enriches wildfly-swarm containers with health checks if the monitoring fraction is present.
+ */
+public class WildFlySwarmHealthCheckEnricher extends AbstractHealthCheckEnricher {
+
+    private static final int DEFAULT_HEALTH_PORT = 8080;
```
Comment (jimmidyson) [inline enricher/fabric8/src/main/java/io/fabric8/maven/enricher/fabric8/WildFlySwarmHealthCheckEnricher.java]: Is there no way we can discover this port? If not, then a config property would be great.

File: enricher/fabric8/src/main/java/io/fabric8/maven/enricher/fabric8/WildFlySwarmHealthCheckEnricher.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+    @Override
+    protected Probe getReadinessProbe() {
+        Probe probe = discoverWildFlySwarmHealthCheck(10);
+        return probe;
+    }
+
+    @Override
+    protected Probe getLivenessProbe() {
+        Probe probe = discoverWildFlySwarmHealthCheck(180);
```
Comment (jimmidyson) [inline enricher/fabric8/src/main/java/io/fabric8/maven/enricher/fabric8/WildFlySwarmHealthCheckEnricher.java]: 180 seconds is quite a long time - do we expect that to be a reasonable startup expectation for swarm containers?

Comment (davsclaus) [inline enricher/fabric8/src/main/java/io/fabric8/maven/enricher/fabric8/WildFlySwarmHealthCheckEnricher.java]: Its using the same values as SB

File: enricher/fabric8/src/main/java/io/fabric8/maven/enricher/fabric8/WildFlySwarmHealthCheckEnricher.java (hunk 1 above)
Comment (davsclaus) [inline enricher/fabric8/src/main/java/io/fabric8/maven/enricher/fabric8/WildFlySwarmHealthCheckEnricher.java]: property from where? In the f-m-p? And if so, how would you do that @rhuss ?

File: enricher/fabric8/src/main/java/io/fabric8/maven/enricher/fabric8/WildFlySwarmHealthCheckEnricher.java (hunk 3 above)
Comment (jimmidyson) [inline enricher/fabric8/src/main/java/io/fabric8/maven/enricher/fabric8/WildFlySwarmHealthCheckEnricher.java]: Yeah OK let's discuss that in a separate issue so we have common settings.

File: enricher/fabric8/src/main/java/io/fabric8/maven/enricher/fabric8/WildFlySwarmHealthCheckEnricher.java (hunk 2 above)
Comment (jimmidyson) [inline enricher/fabric8/src/main/java/io/fabric8/maven/enricher/fabric8/WildFlySwarmHealthCheckEnricher.java]: You can define configurable properties as in https://github.com/fabric8io/fabric8-maven-plugin/blob/master/enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/DependencyEnricher.java#L62-L76 & then retrieve them as in https://github.com/fabric8io/fabric8-maven-plugin/blob/master/enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/DependencyEnricher.java#L172

Comment (davsclaus) [inline enricher/fabric8/src/main/java/io/fabric8/maven/enricher/fabric8/WildFlySwarmHealthCheckEnricher.java]: Thanks @jimmidyson I have just pushed more commits with configurable properties.

**Comment by davsclaus** [comment]
Any thoughts of this?

With this PR then our swarm quickstarts at fabric8 now works and deploy correctly and wont be ready until swarm is ready.

So I would like to see it merge.

Then we can open another ticket about trying to discover wildfly-swarm configuration files and grab port from there, which essentially also the service enricher would need to, if it needs to detect a different port than 8080.

File: enricher/fabric8/src/main/java/io/fabric8/maven/enricher/fabric8/WildFlySwarmHealthCheckEnricher.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+public class WildFlySwarmHealthCheckEnricher extends AbstractHealthCheckEnricher {
+
+    public WildFlySwarmHealthCheckEnricher(EnricherContext buildContext) {
+        super(buildContext, "wildfly-swarm-health-check");
+    }
+
+    // Available configuration keys
+    private enum Config implements Configs.Key {
+
+        defaultScheme {{
```
Comment (jimmidyson) [inline enricher/fabric8/src/main/java/io/fabric8/maven/enricher/fabric8/WildFlySwarmHealthCheckEnricher.java]: As these are actually the names of the properties used, they shouldn't have the `default` prefix AFAIK. @rhuss - halp?

Comment (davsclaus) [inline enricher/fabric8/src/main/java/io/fabric8/maven/enricher/fabric8/WildFlySwarmHealthCheckEnricher.java]: Yeah it makes sense. I pushed a commit without the default prefix.

**Comment by davsclaus** [comment]
[merge]

**Comment by davsclaus** [comment]
Logged a ticket about discovering port number from configuration files
https://github.com/fabric8io/fabric8-maven-plugin/issues/718

**Pull request merged**